### PR TITLE
[askeladd] Per-axis loss autoweighting via running variance

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The binding constraint is wsy and wsz (9.10% and 10.87% vs AB-UPT 3.65%/3.63%). The current loss treats all 5 output axes with fixed weights. Implement **per-axis loss reweighting via running variance**: dynamically scale each axis's loss contribution inversely proportional to its observed running loss variance, auto-tuned from training statistics. This is similar to uncertainty-based weighting (Kendall et al. 2018) but purely empirical: axes with high loss variance get down-weighted (noisy gradient signal), axes with low variance get up-weighted (clean signal). The goal is to let the optimizer spend proportionally more on the shear axes.

## Instructions

Base recipe: 4L/256d, no EMA, Fourier PE, T_max=30, lr=5e-4, surface_weight=2.0, 64 slices.

### Changes to train.py / trainer_runtime.py

1. Add `--autoweight` flag (default False).
2. Maintain a running exponential-moving average of per-axis squared loss: `ema_var[k] = beta * ema_var[k] + (1-beta) * loss_k^2` with `beta=0.99`.
3. Compute per-axis weight: `w_k = 1 / (sqrt(ema_var[k]) + eps)`, then normalise: `w_k /= sum(w_k)`.
4. Apply as a scalar multiplier to each axis's loss before summing (keep `surface_weight=2.0` as a prior multiplicative factor on top of the autoweight for the surface head).
5. Use a **warmup** of 500 steps where weights are uniform (to let the EMA stabilize).

### Axes to weight separately

- `surface_pressure` (ch 0 of surface_preds)
- `wall_shear_x` (ch 1)
- `wall_shear_y` (ch 2) ← binding
- `wall_shear_z` (ch 3) ← binding
- `volume_pressure` (volume_preds ch 0)

### Run

```bash
cd target/ && python train.py \
  --epochs 30 --lr 5e-4 --depth 4 --hidden-dim 256 --num-slices 64 \
  --surface-weight 2.0 --fourier-pe --no-ema \
  --train-surface-points 32768 --train-volume-points 32768 \
  --autoweight \
  --wandb-project DrivAerML --wandb-group askeladd-per-axis-autoweight \
  --kill-thresholds "3000:val_primary/abupt_axis_mean_rel_l2_pct<=25"
```

Log the per-axis weights to W&B at each epoch so we can see whether the optimizer shifts attention toward wsy/wsz.

## Baseline

alphonse Wave 1 run `m9775k1v`: abupt=7.209%, surf_p=4.802%, vol_p=4.166%✓, wsx=7.109%, wsy=9.100%←BINDING, wsz=10.869%←HARDEST

AB-UPT targets (all must be beaten simultaneously):
- surf_p: 3.82%
- vol_p: 6.08% ✓ (already beaten)
- wsx: 5.35%
- wsy: 3.65%
- wsz: 3.63%

### CODE REVIEW

**Comment by morganmcg1** [comment]
**ADVISOR check-in — per-axis loss autoweighting**

Good direction. The per-axis EMA variance weighting is a principled approach to the wsy/wsz gap. A few notes:

**Kill threshold is dead** — the current threshold `3000:val_primary/abupt_axis_mean_rel_l2_pct<=25` fires at step 3000, but the first val event doesn't happen until step ~17,816 (epoch boundary). The threshold will never trigger. Please update to `35000:val_primary/abupt_axis_mean_rel_l2_pct<20` to catch genuine ep2 divergence.

**Log per-axis weights to W&B** — this is essential for debugging. Add a W&B log step that records `w[surf_p]`, `w[wsx]`, `w[wsy]`, `w[wsz]`, `w[vol_p]` at every training step. Without this we can't verify the autoweighting is actually differentiating toward wsy/wsz. The weight trajectory over training is half the scientific value of this experiment.

**500-step warmup correctness** — confirm the uniform warmup is initialized to equal weights (not zeros) and that the EMA denominator is stable before the warmup exits. A NaN in any `ema_var[k]` will silently corrupt all weights via normalization.

**Watch wsy/wsz specifically** — report val wsy and wsz individually in your results comment (not just abupt). Those are the binding axes. Best current: wsy=9.100%, wsz=10.869% (alphonse). Target: wsy=3.65%, wsz=3.63%.

Baseline to beat: val abupt=7.2091% (PR #74, run m9775k1v, 4L/256d Fourier PE, no-EMA, T_max=30).

**Comment by morganmcg1** [comment]
ADVISOR ESCALATION — second check-in.

Askeladd — this is your second advisor message. PR #216 (per-axis EMA variance autoweighting) has been in WIP for over 24 hours with no student response. Your pod shows 1/1 READY.

Please report:
1. Have you launched the training run?
2. If yes: W&B run ID and current epoch
3. If not: implementation blockers or questions

The kill threshold fix (35000:val_primary/abupt_axis_mean_rel_l2_pct<20) from the advisor check-in must also be included. Please confirm.

Action required: respond within the next check-in cycle or the PR will be reassigned.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #216 because ADVISOR (2026-05-01 ~17:35Z): Closing as non-responsive — third strike.

Two prior advisor messages on this PR (initial check-in 14:50Z, escalation 17:19Z) received zero response over ~2.7h. No W&B run ID, no launch confirmation, no diagnostics from askeladd in this window. Pod is 1/1 READY.

Closing this PR per the non-response protocol so the askeladd slot can be reassigned to a fresh hypothesis. The per-axis EMA variance autoweighting idea remains viable and is archived in the Wave 3 ideas pool.

If askeladd is alive and reading this: nothing personal, please respond on your next assignment..
